### ISSUE TITLE
Join Request callback firing with empty data on 32-bit DLL (Unity)

### ISSUE THREAD

**Issue by SirDrMcHurtz** [discordapp/discord-rpc#42]
The game I am working on currently only uses a 32-bit build, but I seem to have an issue with the 32-bit DLL.

When receiving the "join request" callback from another user asking to join, the data packet received is always completely empty. I tried this using the button-clicker example, creating two builds (32-bit and 64-bit) with the 2.0.1 release.

Both builds will fire off the "join request" callback but only the 64-bit will have data.

e.g:
from the logs from the 64-bit version:
**Discord: join request [username]: [userid]**

on the 32-bit version, I just get the basic printout with no values
**Discord: join request :** 

In my actual game the data is also not there with the 32-bit DLL and after trying to use these null values the whole application crashes. It works fine with the 64-bit DLL.

I'm using Unity 2017.1.1.p2

**Comment by crmarsh** [comment]
Weird, I'll check it out

**Comment by SirDrMcHurtz** [comment]
Just to add to this, I tested it and I have the same issue with the latest main release of Unity 2017.2.0.f3.

**Comment by ThomasAunvik** [comment]
I have been having the same problem, the error output is:
```
DllNotFoundException
  at (wrapper managed-to-native) DiscordRpc:RunCallbacks ()
  at DiscordController.Update () [0x00001] in A:\Projects\Project-Moonlit\Moonlit\Assets\Plugins\DiscordRPC\DiscordController.cs:89 
```

**Comment by SirDrMcHurtz** [comment]
@ThomasAunvik Are you running with the 32-Bit DLL? I don't get a DllNotFoundException if I use that, I just get the original bug that I mentioned. 

In my project I made two folders, x86 and x86_64, each with the respective discord-rpc.dll and made sure that each is configured to only be load on 32-bit and 64-bit respectively. 

That will get the DLL loading but I'm still having the issues I mentioned in my original post with the 32-bit DLL (only seems to be the join request that I have issues with)

**Comment by ThomasAunvik** [comment]
Im not sure, what i only did was import it to the project, set the information up, and build it on 32-bit and 64-bit. This is also an error:
`Failed to load 'A:/Projects/Project-Moonlit/Moonlit/Builds/Windows/Moonlit_Data/Plugins/discord-rpc.dll', expected x86 architecture, but was x64 architecture. You must recompile your plugin for x86 architecture.`
Apparently it is a x64 DLL, but i dont know where to get the x86.
Not sure what to do here.
![skjermbilde](https://user-images.githubusercontent.com/9968151/33273775-94fe1868-d38e-11e7-9e5c-b32b86ab1e9a.PNG)

**Comment by SirDrMcHurtz** [comment]
@ThomasAunvik I believe the example Unity project that Discord provided just has the 64-bit DLL included.
You can find the 32-bit DLL under _win32-dynamic\bin\_ in the .zip file

If you try replacing that DLL and do a 32-bit build, you should get it to run.

**Comment by ThomasAunvik** [comment]
Alright, that worked for 32-bit, but not for 64-bit anymore. `Failed to load 'A:/Projects/Project-Moonlit/Moonlit/Builds/64/Moonlit_Data/Plugins/discord-rpc.dll', expected x64 architecture, but was x86 architecture. You must recompile your plugin for x64 architecture.`

**Comment by SirDrMcHurtz** [comment]
Yup you'll need to keep both the DLLs in the project and make sure they are correctly set for 32-bit and 64-bit for the platform (see the platform settings in the screenshot you posted earlier)

The way most projects do it is to have two folders with two DLLs with the same name.

e.g:

**Plugins/Discord/x86/discord-rpc.dll [32-bit only]
Plugins/Discord/x86_64/discord-rpc.dll [64-bit only]**

So get the 32-bit only being used for Windows 32-bit builds and the same for the 64-bit.

**Comment by ThomasAunvik** [comment]
Unfortunately, that did not work.
`Multiple plugins with the same name 'discord-rpc' (found at 'Assets/Plugins/DiscordRPC/x86/discord-rpc.dll' and 'Assets/Plugins/DiscordRPC/x86_64/discord-rpc.dll'). That means one or more plugins are set to be compatible with Editor. Only one plugin at the time can be used by Editor.
`

**Comment by SirDrMcHurtz** [comment]
There are separate settings for the Unity Editor, make sure you disable the 32-bit one for the Editor (assuming you are using 64-bit Unity)

**Comment by ThomasAunvik** [comment]
Alright, that worked for the editor. but it does still not work when building:
```
Plugin 'discord-rpc.dll' is used from several locations:
 Assets/Plugins/DiscordRPC/x86/discord-rpc.dll would be copied to <PluginPath>/discord-rpc.dll
 Assets/Plugins/DiscordRPC/x86_64/discord-rpc.dll would be copied to <PluginPath>/discord-rpc.dll
Please fix plugin settings and try again.

UnityEditor.Modules.DefaultPluginImporterExtension:CheckFileCollisions(String)
UnityEditorInternal.PluginsHelper:CheckFileCollisions(BuildTarget) (at C:/buildslave/unity/build/Editor/Mono/Plugins/PluginsHelper.cs:25)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```
For 32-bit:
![image](https://user-images.githubusercontent.com/9968151/33279154-e1afd7a0-d39d-11e7-9bde-82a7f42f3766.png)


For 64-bit:
![image](https://user-images.githubusercontent.com/9968151/33279119-c4249144-d39d-11e7-84dd-faba8d43b24c.png)
![image](https://user-images.githubusercontent.com/9968151/33279131-ce9f5ab4-d39d-11e7-94b0-1b3c3a2c8dcd.png)

**Comment by SirDrMcHurtz** [comment]
Hmm strange, feel like that should work. That's pretty much exactly the setup I have and it has been building fine. I'd maybe make sure you don't have the DLL hanging around anywhere else, maybe clear up the Unity project temp files and rebuild, sometimes it can get funny with the DLLs.

Not really sure what else to suggest, pretty much any asset I have seen that uses DLLs has them setup like this. It should work!

**Comment by ThomasAunvik** [comment]
Well, quitting unity seems to be removing the temp files already. And i have tried building it then. Still not working. I could try on an empty project.

**Comment by ThomasAunvik** [comment]
Yep, still not working on a different project.
Here is a download link of what the Plugin folder is (Unity Package) https://drive.google.com/file/d/1Tu3LF1r3OiaP52YHshNpJh6hbna5B0ML/view?usp=sharing

**Comment by SirDrMcHurtz** [comment]
That's bizarre. Looks like a weird bug in Unity or something to me. I messed around with that project and had the same issue, but I seem to be able to fix it by configuring it with basically the same settings but inverted.

![image](https://user-images.githubusercontent.com/33718791/33288397-6b8cacc8-d3bc-11e7-9069-dbe0dcdd332d.png)

![image](https://user-images.githubusercontent.com/33718791/33288410-777f3c9e-d3bc-11e7-8895-558a6cc0fce4.png)

Making sure to use the Include Platforms instead, unticking any platform and settings just Windows for x86, and Windows and Editor for x86_64.

I have no idea why this works but it seems to? Unity is weird sometimes...

**Comment by ThomasAunvik** [comment]
Apparently that worked. Thanks.

**Comment by SirDrMcHurtz** [comment]
Any news on this? I still haven't been able to get JoinRequest callback with the 32-bit DLL working at all, it always comes through with null data still. I would love to get the Discord RPC working in my game but this is holding it up.

**Comment by crmarsh** [comment]
I think I have the dlls in the right place in the sample now

**Comment by Fudgecicles** [comment]
Just posting to say that we are also getting the same exact issue, works on 64 bit, doesn't on 32, and it crashes immediately because the request data we get back is empty.

**Comment by msciotti** [comment]
My apologies for not responding much here—I lost track of the discussion and frankly forgot the original bug.

This is...odd, to say the least. I'll have to enlist the help of someone who understands this more than I, but we'll make sure to figure out what's going on.

**Comment by msciotti** [comment]
So, I got my examples in a working state last Friday to attempt this repro. @SirDrMcHurtz, these are the steps I took:

1. Pull the latest repo updates
2. Put the dlls in the right places in the project
3. Build release versions for 32bit, 64bit, and universal windows for safety

In all three versions, I properly received the join request and was able to respond to it without any crashes. Can you confirm if you see the error happening with the same project as well?

**Comment by Fudgecicles** [comment]
Hmmm interesting. I did get the example project working on 32 bit, but the example project in the repo seems a bit messed up - the build helper failed in automatically building, and the actual unity project wouldn't build because BuildHelper.cs wasn't in an editor folder. I just ended up using the latest release for the dlls instead of building them directly. Just in case something different happened there.

But yeah, now I don't really know why this might be happening, as the 32 bit version works in the clicker demo, but even after replacing the dlls in our current project with those dlls it still crashes for the same reason as before. I could provide our builds / source access if that helps? But I don't think we're doing anything that the clicker demo is not, so I'm not sure why things would be different.

**Comment by Fudgecicles** [comment]
Added bit a weirdness, I'm pretty sure all the other discord features in our 32 bit build are working, so its definitely loading and running the dll correctly.

**Comment by SirDrMcHurtz** [comment]
@msciotti Interesting! Would it be possible to provide the example project you are using with your DLL structure, too see if I stop getting the issue as well? Also what version of Unity are you using out of interest?

**Comment by msciotti** [comment]
@Fudgecicles: With regards to the `BuildHelpers.cs` thing...yeah. It's not properly `IF`d at runtime. I ended up needing to comment it all out when doing build and run. It needs to be fixed.

@SirDrMcHurtz: I'm using the `button-clicker` example in this repo. The DLL structure is...standard? I put the 32bit dll in `Assets\Plugins\x86`. Unity version is `2017.1.0f3 Personal`.

**Comment by Fudgecicles** [comment]
Ok, I think I figured it out by looking at differences between the repo `DiscordController.cs` and our local one. 

The only difference was that the definition for `RequestCallback(DiscordRpc.JoinRequest request)` was declared as `ref` on the remote version. Adding this to mine fixed the problem. Not sure how calling unmanaged code works exactly, but maybe caused by a shallow copy... or something? I don't get it. Though I don't understand why it would work on 64 bit and not 32 bit systems then. Either way, we are able to accept join requests on 32 bit builds now so seems fixed to me.

**Comment by msciotti** [comment]
Awesome. One down, one to go! @SirDrMcHurtz if you can confirm that the sample app works properly (and perhaps take a look at the `ref` thing in your own code), I'll mark this issue as resolved.

**Comment by SirDrMcHurtz** [comment]
Ok, I'm very confused by this as I'm 90% sure that I checked that exact fix a couple of weeks ago (in fact it's still in my codebase) but it didn't seem to fix it.

But to be double sure I checked again today and it seems to work now? Maybe it's with the most recent DLL? Either way, I'm a bit confused but it does seem to be working now with my 32-bit builds, so yeah feel free to close this issue.
